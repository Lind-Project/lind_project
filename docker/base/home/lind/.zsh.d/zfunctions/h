## Fix zsh annoying history behavior
 if [[ -z "$*" ]]; then history; else egrep "$*" <(fc -l 1); fi
