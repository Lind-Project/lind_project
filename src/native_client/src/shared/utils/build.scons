# -*- python2 -*-
# Copyright 2008 The Native Client Authors.  All rights reserved.
# Use of this source code is governed by a BSD-style license that can
# be found in the LICENSE file.

import os
import sys
Import('env')

env.ComponentLibrary('utils',
                     ['flags.c',
                      'formatting.c'])
