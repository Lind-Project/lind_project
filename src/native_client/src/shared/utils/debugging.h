/*
 * Copyright 2009 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can
 * be found in the LICENSE file.
 */

/* Define simple debugging utilities that are turned on/off by the
 * value of the define flag DEBUGGING.
 *
 * To turn on debugging of instruction decoding, change value of
 * DEBUGGING to 1.
 *
 */

#ifndef NATIVE_CLIENT_SRC_SHARED_UTILS_DEBUGGING_H__
#define NATIVE_CLIENT_SRC_SHARED_UTILS_DEBUGGING_H__

/* Turn off debugging if not otherwise specified in the specific code file. */
#ifndef DEBUGGING
#define DEBUGGING 0
#endif

#if DEBUGGING
/* Defines to execute statement(s) s if in DEBUGGING mode, and compile
 * in either mode. This allows type checking to be applied at all times.
 */
#define DEBUG(s) s
#else
/* Defines to compile but not include statement(s) s if not
 * in DEBUGGING mode.
 */
#define DEBUG(s) do { if (0) { s; } } while (0)
#endif

#if DEBUGGING
/* Defines to compile execute statement(s) if in DEBUGGING mode.
 * Erases otherwise. This should only be used when linking
 * would otherwise require the symbol when DEBUGGING is false.
*/
#define DEBUG_OR_ERASE(s) s
#else
/* Defines to erase s if not in DEBUGGING mode. */
#define DEBUG_OR_ERASE(s)
#endif

#endif  /* NATIVE_CLIENT_SRC_SHARED_UTILS_DEBUGGING_H__ */
