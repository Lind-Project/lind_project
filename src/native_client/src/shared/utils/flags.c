/*
 * Copyright 2009 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can
 * be found in the LICENSE file.
 */

/*
 * Some useful routines to handle command line arguments.
 */

#include <stdlib.h>
#include <string.h>

#include "native_client/src/shared/utils/flags.h"

/*
 * Given a pointer to a flag name to be
 * checked against the pointer to the corresponding
 * argument, return true iff the argument begins with
 * the flag name. During the match, advance both pointers
 * to the maximal matched prefix.
 */
static Bool MatchFlagPrefix(const char** name,
                            const char** arg) {
  while (*(*name)) {
    if (*(*name) != *(*arg)) {
      return FALSE;
    }
    ++(*name);
    ++(*arg);
  }
  return TRUE;
}

/*
 * Given a pointer to a flag name to be checked
 * against the pointer to the corresponding argument, return true
 * iff the argument begins with the flag name, followed by
 * and equals sign. During the match, advance both pointers
 * to the maximal matched prefix.
 */
static Bool MatchFlagPrefixEquals(const char** name,
                                  const char** arg) {
  const char* equals = "=";
  return MatchFlagPrefix(name, arg) &&
      MatchFlagPrefix(&equals, arg);
}

Bool GrokBoolFlag(const char* name,
                  const char* arg,
                  Bool* flag) {
  if (strcmp(name, arg) == 0) {
    *flag = TRUE;
    return TRUE;
  } else if (MatchFlagPrefixEquals(&name, &arg)) {
    if (strcmp(arg, "true") == 0) {
      *flag = TRUE;
      return TRUE;
    } else if (strcmp(arg, "false") == 0) {
      *flag = FALSE;
      return TRUE;
    }
  }
  return FALSE;
}

static Bool IsZero(const char* arg) {
  while (*arg) {
    if ('0' != *arg) {
      return FALSE;
    }
    ++arg;
  }
  return TRUE;
}

Bool GrokIntFlag(const char* name,
                 const char* arg,
                 int* flag) {
  if (MatchFlagPrefixEquals(&name, &arg)) {
    int value = atoi(arg);
    /* verify that arg is all zeros. Otherwise,
     * assume that 0 was returned due to an error.
     */
    if (0 == value && !IsZero(arg)) return FALSE;
    *flag = value;
    return TRUE;
  }
  return FALSE;
}

Bool GrokUint32HexFlag(const char* name,
                       const char* arg,
                       uint32_t* flag) {
  if (MatchFlagPrefixEquals(&name, &arg)) {
    unsigned long value = strtoul(arg, NULL, 16);
    /* Verify that arg is all zeros. Otherwise,
     * assume that 0 was returned due to an error.
     */
    if (0L == value && !IsZero(arg)) return FALSE;
    *flag = (uint32_t) value;
    return TRUE;
  }
  return FALSE;
}


Bool GrokCstringFlag(const char* name,
                     const char* arg,
                     char** flag) {
  if (MatchFlagPrefixEquals(&name, &arg)) {
    *flag = (char*) arg;
    return TRUE;
  }
  return FALSE;
}
