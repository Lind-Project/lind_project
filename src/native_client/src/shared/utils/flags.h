/*
 * Copyright 2009 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can
 * be found in the LICENSE file.
 */

/*
 * Some useful routines to handle command line arguments.
 */

#ifndef NATIVE_CLIENT_SRC_SHARED_UTILS_FLAGS_H__
#define NATIVE_CLIENT_SRC_SHARED_UTILS_FLAGS_H__

#include "native_client/src/shared/utils/types.h"

EXTERN_C_BEGIN

/*
 * Attempts to parse a boolean command line argument.
 * Returns TRUE iff able to parse and set the flag.
 *
 * name - The (command line) flag name.
 * arg - The command line argument to process.
 * flag - Pointer to the corresponding boolean flag.
 */
Bool GrokBoolFlag(const char* name,
                  const char* arg,
                  Bool* flag);

/*
 * Attempts to parse an integer command line argment.
 * Returns TRUE iff able to parse and set the flag.
 *
 * name - The (command line) flag name.
 * arg - The command line argument to process.
 * flag - Pointer to the corresponding integer flag.
 */
Bool GrokIntFlag(const char* name,
                 const char* arg,
                 int* flag);

/*
 * Attempts to parse a char* command line argument.
 * Returns TRUE iff able to parse and set the flag.
 *
 * name - The (command line) flag name.
 * arg - The command line argument to process.
 * flag - Pointer to the corresponding char* flag.
 */
Bool GrokCstringFlag(const char* name,
                     const char* arg,
                     char** flag);

/*
 * Attempts to parse a 32 bit hex valued flag.
 * Returns TRUE iff able to parse and set the flag.
 *
 * name - The (command line) flag name.
 * arg - The command line argument to process.
 * flag - Pointer to the corresponding uint32_t flag.
 */
Bool GrokUint32HexFlag(const char* name,
                       const char* arg,
                       uint32_t* flag);

EXTERN_C_END

#endif  /*  NATIVE_CLIENT_SRC_SHARED_UTILS_FLAGS_H__ */
