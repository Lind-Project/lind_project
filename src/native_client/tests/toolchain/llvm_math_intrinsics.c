/*
 * Copyright (c) 2011 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * This test ensures that the front/middle/backends can deal with the
 * math.h functions which have builtins and llvm intrinsics.
 *
 * There are two categories.
 *
 * (1) Standard uses without resorting to llvm assembly:
 * direct calls to C library function, or the gcc __builtin_sin(), etc.
 *
 * (2) Directly testing llvm.* intrinsics. These we test when we know
 * that they *will* be in bitcode or are useful (e.g., we know hardware
 * acceleration is available).  We currently use "llvm.sqrt.*" within
 * the libm sqrt() function: newlib-trunk/newlib/libm/machine/pnacl/w_sqrt.c
 * The libm sqrt is coded to have error checking and errno setting external
 * to the llvm.sqrt.* call.  Other uses of llvm.sqrt have no guarantees
 * about setting errno. This tests that we don't get infinite recursion
 * from such usage (e.g., if a backend expands llvm.sqrt.f64 within sqrt()
 * back into a call to sqrt()).
 */

#include <math.h>
#include <stdio.h>
#include "native_client/tests/toolchain/utils.h"

/* Volatile to prevent library-call constant folding optimizations. */
volatile float f32[] =  {NAN, -INFINITY, -HUGE_VALF,
                         /* e^(-M_PI) may be broken on mips QEMU. */
                         -0.5, -0.0, 0.0,
                         5.0, 16.0, 10.0, M_PI, M_PI_2, M_E,
                         HUGE_VALF, INFINITY };
volatile double f64[] = {NAN, -INFINITY, -HUGE_VAL,
                         /* e^(-M_PI) may be broken on mips QEMU. */
                         -0.5, -0.0, 0.0,
                         5.0, 16.0, 10.0, M_PI, M_PI_2, M_E,
                         HUGE_VAL, INFINITY };

volatile float base32 = 2.0;
volatile float neg_base32 = -2.0;
volatile double base64 = 2.0;
volatile double neg_base64 = -2.0;

/*
 * The LLVM language reference considers this undefined for values < -0.0.
 * In practice hardware sqrt instructions returns NaN in that case.
 * We will need to guarantee this behavior.
 */
double llvm_intrinsic_sqrt(double) __asm__("llvm.sqrt.f64");
float llvm_intrinsic_sqrtf(float) __asm__("llvm.sqrt.f32");

/*
 * Note: This can end up printing NAN and INFINITY values.
 * For floating point conversions to strings, C99 only asks for the
 * prefix to match "nan" and "inf".  The case can vary as well.
 * Watch out for standard library changes.
 */
#define print_op1(prec, op, x)                                      \
  printf("%s (math.h): %." #prec "f\n", #op, op(x));                \
  printf("%s (builtin): %." #prec "f\n", #op, __builtin_ ## op(x));

#define print_op1_llvm(prec, op, x)                                     \
  printf("%s (llvm): %." #prec "f\n", #op, llvm_intrinsic_ ## op(x))

#define print_op2(prec, op, x, y)                                       \
  printf("%s(%f, %f) (math.h): %." #prec "f\n", #op, x, y, op(x, y));   \
  printf("%s(%f, %f) (builtin): %." #prec "f\n", #op, \
         x, y, __builtin_ ## op(x, y));

int main(int argc, char* argv[]) {
  int i;

  /*
   * Sqrt can end up converting SNaN to infinity under MIPS QEMU 0.12.
   * For now, use 0.0f / 0.0f to work around that.
   * http://code.google.com/p/nativeclient/issues/detail?id=3533
   *
   * Replace the NAN entries.
   */
  volatile float zero = 0.0f; f32[0] = zero / zero; f64[0] = zero / zero;

  for (i = 0; i < ARRAY_SIZE_UNSAFE(f32); ++i) {
    printf("\nf32 value is: %.6f\n",  f32[i]);
    print_op1(5, sqrtf, f32[i]);
    print_op1_llvm(5, sqrtf, f32[i]);
    print_op1(5, logf, f32[i]);
    print_op1(5, log2f, f32[i]);
    print_op1(5, log10f, f32[i]);
    print_op1(4, expf, f32[i]);
    print_op1(4, exp2f, f32[i]);
    print_op1(5, sinf, f32[i]);
    print_op1(5, cosf, f32[i]);
    print_op2(5, powf, neg_base32, f32[i]);
    print_op2(5, powf, base32, f32[i]);
  }

  for (i = 0; i < ARRAY_SIZE_UNSAFE(f64); ++i) {
    printf("\nf64 value is: %.6f\n",  f64[i]);
    print_op1(6, sqrt, f64[i]);
    print_op1_llvm(6, sqrt, f64[i]);
    print_op1(6, log, f64[i]);
    print_op1(6, log2, f64[i]);
    print_op1(6, log10, f64[i]);
    print_op1(6, exp, f64[i]);
    print_op1(6, exp2, f64[i]);
    print_op1(6, sin, f64[i]);
    print_op1(6, cos, f64[i]);
    print_op2(6, pow, neg_base64, f64[i]);
    print_op2(6, pow, base64, f64[i]);
  }

  return 0;
}
