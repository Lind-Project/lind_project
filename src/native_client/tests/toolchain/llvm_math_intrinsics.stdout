
f32 value is: nan
sqrtf (math.h): nan
sqrtf (builtin): nan
sqrtf (llvm): nan
logf (math.h): nan
logf (builtin): nan
log2f (math.h): nan
log2f (builtin): nan
log10f (math.h): nan
log10f (builtin): nan
expf (math.h): nan
expf (builtin): nan
exp2f (math.h): nan
exp2f (builtin): nan
sinf (math.h): nan
sinf (builtin): nan
cosf (math.h): nan
cosf (builtin): nan
powf(-2.000000, nan) (math.h): nan
powf(-2.000000, nan) (builtin): nan
powf(2.000000, nan) (math.h): nan
powf(2.000000, nan) (builtin): nan

f32 value is: -inf
sqrtf (math.h): nan
sqrtf (builtin): nan
sqrtf (llvm): nan
logf (math.h): nan
logf (builtin): nan
log2f (math.h): nan
log2f (builtin): nan
log10f (math.h): nan
log10f (builtin): nan
expf (math.h): 0.0000
expf (builtin): 0.0000
exp2f (math.h): 0.0000
exp2f (builtin): 0.0000
sinf (math.h): nan
sinf (builtin): nan
cosf (math.h): nan
cosf (builtin): nan
powf(-2.000000, -inf) (math.h): 0.00000
powf(-2.000000, -inf) (builtin): 0.00000
powf(2.000000, -inf) (math.h): 0.00000
powf(2.000000, -inf) (builtin): 0.00000

f32 value is: -inf
sqrtf (math.h): nan
sqrtf (builtin): nan
sqrtf (llvm): nan
logf (math.h): nan
logf (builtin): nan
log2f (math.h): nan
log2f (builtin): nan
log10f (math.h): nan
log10f (builtin): nan
expf (math.h): 0.0000
expf (builtin): 0.0000
exp2f (math.h): 0.0000
exp2f (builtin): 0.0000
sinf (math.h): nan
sinf (builtin): nan
cosf (math.h): nan
cosf (builtin): nan
powf(-2.000000, -inf) (math.h): 0.00000
powf(-2.000000, -inf) (builtin): 0.00000
powf(2.000000, -inf) (math.h): 0.00000
powf(2.000000, -inf) (builtin): 0.00000

f32 value is: -0.500000
sqrtf (math.h): nan
sqrtf (builtin): nan
sqrtf (llvm): nan
logf (math.h): nan
logf (builtin): nan
log2f (math.h): nan
log2f (builtin): nan
log10f (math.h): nan
log10f (builtin): nan
expf (math.h): 0.6065
expf (builtin): 0.6065
exp2f (math.h): 0.7071
exp2f (builtin): 0.7071
sinf (math.h): -0.47943
sinf (builtin): -0.47943
cosf (math.h): 0.87758
cosf (builtin): 0.87758
powf(-2.000000, -0.500000) (math.h): nan
powf(-2.000000, -0.500000) (builtin): nan
powf(2.000000, -0.500000) (math.h): 0.70711
powf(2.000000, -0.500000) (builtin): 0.70711

f32 value is: -0.000000
sqrtf (math.h): -0.00000
sqrtf (builtin): -0.00000
sqrtf (llvm): -0.00000
logf (math.h): -inf
logf (builtin): -inf
log2f (math.h): -inf
log2f (builtin): -inf
log10f (math.h): -inf
log10f (builtin): -inf
expf (math.h): 1.0000
expf (builtin): 1.0000
exp2f (math.h): 1.0000
exp2f (builtin): 1.0000
sinf (math.h): -0.00000
sinf (builtin): -0.00000
cosf (math.h): 1.00000
cosf (builtin): 1.00000
powf(-2.000000, -0.000000) (math.h): 1.00000
powf(-2.000000, -0.000000) (builtin): 1.00000
powf(2.000000, -0.000000) (math.h): 1.00000
powf(2.000000, -0.000000) (builtin): 1.00000

f32 value is: 0.000000
sqrtf (math.h): 0.00000
sqrtf (builtin): 0.00000
sqrtf (llvm): 0.00000
logf (math.h): -inf
logf (builtin): -inf
log2f (math.h): -inf
log2f (builtin): -inf
log10f (math.h): -inf
log10f (builtin): -inf
expf (math.h): 1.0000
expf (builtin): 1.0000
exp2f (math.h): 1.0000
exp2f (builtin): 1.0000
sinf (math.h): 0.00000
sinf (builtin): 0.00000
cosf (math.h): 1.00000
cosf (builtin): 1.00000
powf(-2.000000, 0.000000) (math.h): 1.00000
powf(-2.000000, 0.000000) (builtin): 1.00000
powf(2.000000, 0.000000) (math.h): 1.00000
powf(2.000000, 0.000000) (builtin): 1.00000

f32 value is: 5.000000
sqrtf (math.h): 2.23607
sqrtf (builtin): 2.23607
sqrtf (llvm): 2.23607
logf (math.h): 1.60944
logf (builtin): 1.60944
log2f (math.h): 2.32193
log2f (builtin): 2.32193
log10f (math.h): 0.69897
log10f (builtin): 0.69897
expf (math.h): 148.4132
expf (builtin): 148.4132
exp2f (math.h): 32.0000
exp2f (builtin): 32.0000
sinf (math.h): -0.95892
sinf (builtin): -0.95892
cosf (math.h): 0.28366
cosf (builtin): 0.28366
powf(-2.000000, 5.000000) (math.h): -32.00000
powf(-2.000000, 5.000000) (builtin): -32.00000
powf(2.000000, 5.000000) (math.h): 32.00000
powf(2.000000, 5.000000) (builtin): 32.00000

f32 value is: 16.000000
sqrtf (math.h): 4.00000
sqrtf (builtin): 4.00000
sqrtf (llvm): 4.00000
logf (math.h): 2.77259
logf (builtin): 2.77259
log2f (math.h): 4.00000
log2f (builtin): 4.00000
log10f (math.h): 1.20412
log10f (builtin): 1.20412
expf (math.h): 8886111.0000
expf (builtin): 8886111.0000
exp2f (math.h): 65536.0000
exp2f (builtin): 65536.0000
sinf (math.h): -0.28790
sinf (builtin): -0.28790
cosf (math.h): -0.95766
cosf (builtin): -0.95766
powf(-2.000000, 16.000000) (math.h): 65536.00000
powf(-2.000000, 16.000000) (builtin): 65536.00000
powf(2.000000, 16.000000) (math.h): 65536.00000
powf(2.000000, 16.000000) (builtin): 65536.00000

f32 value is: 10.000000
sqrtf (math.h): 3.16228
sqrtf (builtin): 3.16228
sqrtf (llvm): 3.16228
logf (math.h): 2.30259
logf (builtin): 2.30259
log2f (math.h): 3.32193
log2f (builtin): 3.32193
log10f (math.h): 1.00000
log10f (builtin): 1.00000
expf (math.h): 22026.4648
expf (builtin): 22026.4648
exp2f (math.h): 1024.0000
exp2f (builtin): 1024.0000
sinf (math.h): -0.54402
sinf (builtin): -0.54402
cosf (math.h): -0.83907
cosf (builtin): -0.83907
powf(-2.000000, 10.000000) (math.h): 1024.00000
powf(-2.000000, 10.000000) (builtin): 1024.00000
powf(2.000000, 10.000000) (math.h): 1024.00000
powf(2.000000, 10.000000) (builtin): 1024.00000

f32 value is: 3.141593
sqrtf (math.h): 1.77245
sqrtf (builtin): 1.77245
sqrtf (llvm): 1.77245
logf (math.h): 1.14473
logf (builtin): 1.14473
log2f (math.h): 1.65150
log2f (builtin): 1.65150
log10f (math.h): 0.49715
log10f (builtin): 0.49715
expf (math.h): 23.1407
expf (builtin): 23.1407
exp2f (math.h): 8.8250
exp2f (builtin): 8.8250
sinf (math.h): -0.00000
sinf (builtin): -0.00000
cosf (math.h): -1.00000
cosf (builtin): -1.00000
powf(-2.000000, 3.141593) (math.h): nan
powf(-2.000000, 3.141593) (builtin): nan
powf(2.000000, 3.141593) (math.h): 8.82498
powf(2.000000, 3.141593) (builtin): 8.82498

f32 value is: 1.570796
sqrtf (math.h): 1.25331
sqrtf (builtin): 1.25331
sqrtf (llvm): 1.25331
logf (math.h): 0.45158
logf (builtin): 0.45158
log2f (math.h): 0.65150
log2f (builtin): 0.65150
log10f (math.h): 0.19612
log10f (builtin): 0.19612
expf (math.h): 4.8105
expf (builtin): 4.8105
exp2f (math.h): 2.9707
exp2f (builtin): 2.9707
sinf (math.h): 1.00000
sinf (builtin): 1.00000
cosf (math.h): -0.00000
cosf (builtin): -0.00000
powf(-2.000000, 1.570796) (math.h): nan
powf(-2.000000, 1.570796) (builtin): nan
powf(2.000000, 1.570796) (math.h): 2.97069
powf(2.000000, 1.570796) (builtin): 2.97069

f32 value is: 2.718282
sqrtf (math.h): 1.64872
sqrtf (builtin): 1.64872
sqrtf (llvm): 1.64872
logf (math.h): 1.00000
logf (builtin): 1.00000
log2f (math.h): 1.44270
log2f (builtin): 1.44270
log10f (math.h): 0.43429
log10f (builtin): 0.43429
expf (math.h): 15.1543
expf (builtin): 15.1543
exp2f (math.h): 6.5809
exp2f (builtin): 6.5809
sinf (math.h): 0.41078
sinf (builtin): 0.41078
cosf (math.h): -0.91173
cosf (builtin): -0.91173
powf(-2.000000, 2.718282) (math.h): nan
powf(-2.000000, 2.718282) (builtin): nan
powf(2.000000, 2.718282) (math.h): 6.58089
powf(2.000000, 2.718282) (builtin): 6.58089

f32 value is: inf
sqrtf (math.h): inf
sqrtf (builtin): inf
sqrtf (llvm): inf
logf (math.h): inf
logf (builtin): inf
log2f (math.h): inf
log2f (builtin): inf
log10f (math.h): inf
log10f (builtin): inf
expf (math.h): inf
expf (builtin): inf
exp2f (math.h): inf
exp2f (builtin): inf
sinf (math.h): nan
sinf (builtin): nan
cosf (math.h): nan
cosf (builtin): nan
powf(-2.000000, inf) (math.h): inf
powf(-2.000000, inf) (builtin): inf
powf(2.000000, inf) (math.h): inf
powf(2.000000, inf) (builtin): inf

f32 value is: inf
sqrtf (math.h): inf
sqrtf (builtin): inf
sqrtf (llvm): inf
logf (math.h): inf
logf (builtin): inf
log2f (math.h): inf
log2f (builtin): inf
log10f (math.h): inf
log10f (builtin): inf
expf (math.h): inf
expf (builtin): inf
exp2f (math.h): inf
exp2f (builtin): inf
sinf (math.h): nan
sinf (builtin): nan
cosf (math.h): nan
cosf (builtin): nan
powf(-2.000000, inf) (math.h): inf
powf(-2.000000, inf) (builtin): inf
powf(2.000000, inf) (math.h): inf
powf(2.000000, inf) (builtin): inf

f64 value is: nan
sqrt (math.h): nan
sqrt (builtin): nan
sqrt (llvm): nan
log (math.h): nan
log (builtin): nan
log2 (math.h): nan
log2 (builtin): nan
log10 (math.h): nan
log10 (builtin): nan
exp (math.h): nan
exp (builtin): nan
exp2 (math.h): nan
exp2 (builtin): nan
sin (math.h): nan
sin (builtin): nan
cos (math.h): nan
cos (builtin): nan
pow(-2.000000, nan) (math.h): nan
pow(-2.000000, nan) (builtin): nan
pow(2.000000, nan) (math.h): nan
pow(2.000000, nan) (builtin): nan

f64 value is: -inf
sqrt (math.h): nan
sqrt (builtin): nan
sqrt (llvm): nan
log (math.h): nan
log (builtin): nan
log2 (math.h): nan
log2 (builtin): nan
log10 (math.h): nan
log10 (builtin): nan
exp (math.h): 0.000000
exp (builtin): 0.000000
exp2 (math.h): 0.000000
exp2 (builtin): 0.000000
sin (math.h): nan
sin (builtin): nan
cos (math.h): nan
cos (builtin): nan
pow(-2.000000, -inf) (math.h): 0.000000
pow(-2.000000, -inf) (builtin): 0.000000
pow(2.000000, -inf) (math.h): 0.000000
pow(2.000000, -inf) (builtin): 0.000000

f64 value is: -inf
sqrt (math.h): nan
sqrt (builtin): nan
sqrt (llvm): nan
log (math.h): nan
log (builtin): nan
log2 (math.h): nan
log2 (builtin): nan
log10 (math.h): nan
log10 (builtin): nan
exp (math.h): 0.000000
exp (builtin): 0.000000
exp2 (math.h): 0.000000
exp2 (builtin): 0.000000
sin (math.h): nan
sin (builtin): nan
cos (math.h): nan
cos (builtin): nan
pow(-2.000000, -inf) (math.h): 0.000000
pow(-2.000000, -inf) (builtin): 0.000000
pow(2.000000, -inf) (math.h): 0.000000
pow(2.000000, -inf) (builtin): 0.000000

f64 value is: -0.500000
sqrt (math.h): nan
sqrt (builtin): nan
sqrt (llvm): nan
log (math.h): nan
log (builtin): nan
log2 (math.h): nan
log2 (builtin): nan
log10 (math.h): nan
log10 (builtin): nan
exp (math.h): 0.606531
exp (builtin): 0.606531
exp2 (math.h): 0.707107
exp2 (builtin): 0.707107
sin (math.h): -0.479426
sin (builtin): -0.479426
cos (math.h): 0.877583
cos (builtin): 0.877583
pow(-2.000000, -0.500000) (math.h): nan
pow(-2.000000, -0.500000) (builtin): nan
pow(2.000000, -0.500000) (math.h): 0.707107
pow(2.000000, -0.500000) (builtin): 0.707107

f64 value is: -0.000000
sqrt (math.h): -0.000000
sqrt (builtin): -0.000000
sqrt (llvm): -0.000000
log (math.h): -inf
log (builtin): -inf
log2 (math.h): -inf
log2 (builtin): -inf
log10 (math.h): -inf
log10 (builtin): -inf
exp (math.h): 1.000000
exp (builtin): 1.000000
exp2 (math.h): 1.000000
exp2 (builtin): 1.000000
sin (math.h): -0.000000
sin (builtin): -0.000000
cos (math.h): 1.000000
cos (builtin): 1.000000
pow(-2.000000, -0.000000) (math.h): 1.000000
pow(-2.000000, -0.000000) (builtin): 1.000000
pow(2.000000, -0.000000) (math.h): 1.000000
pow(2.000000, -0.000000) (builtin): 1.000000

f64 value is: 0.000000
sqrt (math.h): 0.000000
sqrt (builtin): 0.000000
sqrt (llvm): 0.000000
log (math.h): -inf
log (builtin): -inf
log2 (math.h): -inf
log2 (builtin): -inf
log10 (math.h): -inf
log10 (builtin): -inf
exp (math.h): 1.000000
exp (builtin): 1.000000
exp2 (math.h): 1.000000
exp2 (builtin): 1.000000
sin (math.h): 0.000000
sin (builtin): 0.000000
cos (math.h): 1.000000
cos (builtin): 1.000000
pow(-2.000000, 0.000000) (math.h): 1.000000
pow(-2.000000, 0.000000) (builtin): 1.000000
pow(2.000000, 0.000000) (math.h): 1.000000
pow(2.000000, 0.000000) (builtin): 1.000000

f64 value is: 5.000000
sqrt (math.h): 2.236068
sqrt (builtin): 2.236068
sqrt (llvm): 2.236068
log (math.h): 1.609438
log (builtin): 1.609438
log2 (math.h): 2.321928
log2 (builtin): 2.321928
log10 (math.h): 0.698970
log10 (builtin): 0.698970
exp (math.h): 148.413159
exp (builtin): 148.413159
exp2 (math.h): 32.000000
exp2 (builtin): 32.000000
sin (math.h): -0.958924
sin (builtin): -0.958924
cos (math.h): 0.283662
cos (builtin): 0.283662
pow(-2.000000, 5.000000) (math.h): -32.000000
pow(-2.000000, 5.000000) (builtin): -32.000000
pow(2.000000, 5.000000) (math.h): 32.000000
pow(2.000000, 5.000000) (builtin): 32.000000

f64 value is: 16.000000
sqrt (math.h): 4.000000
sqrt (builtin): 4.000000
sqrt (llvm): 4.000000
log (math.h): 2.772589
log (builtin): 2.772589
log2 (math.h): 4.000000
log2 (builtin): 4.000000
log10 (math.h): 1.204120
log10 (builtin): 1.204120
exp (math.h): 8886110.520508
exp (builtin): 8886110.520508
exp2 (math.h): 65536.000000
exp2 (builtin): 65536.000000
sin (math.h): -0.287903
sin (builtin): -0.287903
cos (math.h): -0.957659
cos (builtin): -0.957659
pow(-2.000000, 16.000000) (math.h): 65536.000000
pow(-2.000000, 16.000000) (builtin): 65536.000000
pow(2.000000, 16.000000) (math.h): 65536.000000
pow(2.000000, 16.000000) (builtin): 65536.000000

f64 value is: 10.000000
sqrt (math.h): 3.162278
sqrt (builtin): 3.162278
sqrt (llvm): 3.162278
log (math.h): 2.302585
log (builtin): 2.302585
log2 (math.h): 3.321928
log2 (builtin): 3.321928
log10 (math.h): 1.000000
log10 (builtin): 1.000000
exp (math.h): 22026.465795
exp (builtin): 22026.465795
exp2 (math.h): 1024.000000
exp2 (builtin): 1024.000000
sin (math.h): -0.544021
sin (builtin): -0.544021
cos (math.h): -0.839072
cos (builtin): -0.839072
pow(-2.000000, 10.000000) (math.h): 1024.000000
pow(-2.000000, 10.000000) (builtin): 1024.000000
pow(2.000000, 10.000000) (math.h): 1024.000000
pow(2.000000, 10.000000) (builtin): 1024.000000

f64 value is: 3.141593
sqrt (math.h): 1.772454
sqrt (builtin): 1.772454
sqrt (llvm): 1.772454
log (math.h): 1.144730
log (builtin): 1.144730
log2 (math.h): 1.651496
log2 (builtin): 1.651496
log10 (math.h): 0.497150
log10 (builtin): 0.497150
exp (math.h): 23.140693
exp (builtin): 23.140693
exp2 (math.h): 8.824978
exp2 (builtin): 8.824978
sin (math.h): 0.000000
sin (builtin): 0.000000
cos (math.h): -1.000000
cos (builtin): -1.000000
pow(-2.000000, 3.141593) (math.h): nan
pow(-2.000000, 3.141593) (builtin): nan
pow(2.000000, 3.141593) (math.h): 8.824978
pow(2.000000, 3.141593) (builtin): 8.824978

f64 value is: 1.570796
sqrt (math.h): 1.253314
sqrt (builtin): 1.253314
sqrt (llvm): 1.253314
log (math.h): 0.451583
log (builtin): 0.451583
log2 (math.h): 0.651496
log2 (builtin): 0.651496
log10 (math.h): 0.196120
log10 (builtin): 0.196120
exp (math.h): 4.810477
exp (builtin): 4.810477
exp2 (math.h): 2.970686
exp2 (builtin): 2.970686
sin (math.h): 1.000000
sin (builtin): 1.000000
cos (math.h): 0.000000
cos (builtin): 0.000000
pow(-2.000000, 1.570796) (math.h): nan
pow(-2.000000, 1.570796) (builtin): nan
pow(2.000000, 1.570796) (math.h): 2.970686
pow(2.000000, 1.570796) (builtin): 2.970686

f64 value is: 2.718282
sqrt (math.h): 1.648721
sqrt (builtin): 1.648721
sqrt (llvm): 1.648721
log (math.h): 1.000000
log (builtin): 1.000000
log2 (math.h): 1.442695
log2 (builtin): 1.442695
log10 (math.h): 0.434294
log10 (builtin): 0.434294
exp (math.h): 15.154262
exp (builtin): 15.154262
exp2 (math.h): 6.580886
exp2 (builtin): 6.580886
sin (math.h): 0.410781
sin (builtin): 0.410781
cos (math.h): -0.911734
cos (builtin): -0.911734
pow(-2.000000, 2.718282) (math.h): nan
pow(-2.000000, 2.718282) (builtin): nan
pow(2.000000, 2.718282) (math.h): 6.580886
pow(2.000000, 2.718282) (builtin): 6.580886

f64 value is: inf
sqrt (math.h): inf
sqrt (builtin): inf
sqrt (llvm): inf
log (math.h): inf
log (builtin): inf
log2 (math.h): inf
log2 (builtin): inf
log10 (math.h): inf
log10 (builtin): inf
exp (math.h): inf
exp (builtin): inf
exp2 (math.h): inf
exp2 (builtin): inf
sin (math.h): nan
sin (builtin): nan
cos (math.h): nan
cos (builtin): nan
pow(-2.000000, inf) (math.h): inf
pow(-2.000000, inf) (builtin): inf
pow(2.000000, inf) (math.h): inf
pow(2.000000, inf) (builtin): inf

f64 value is: inf
sqrt (math.h): inf
sqrt (builtin): inf
sqrt (llvm): inf
log (math.h): inf
log (builtin): inf
log2 (math.h): inf
log2 (builtin): inf
log10 (math.h): inf
log10 (builtin): inf
exp (math.h): inf
exp (builtin): inf
exp2 (math.h): inf
exp2 (builtin): inf
sin (math.h): nan
sin (builtin): nan
cos (math.h): nan
cos (builtin): nan
pow(-2.000000, inf) (math.h): inf
pow(-2.000000, inf) (builtin): inf
pow(2.000000, inf) (math.h): inf
pow(2.000000, inf) (builtin): inf
